### PR TITLE
Product Subscriptions: Fix crash for `SubscriptionExpiryViewModel`

### PR DESCRIPTION
**Pull request by itsmeichigo**
Closes: #11458 

## Description

On`SubscriptionExpiryViewModel` we are using `stride` to create a sequence from 2 values with steps defined by `periodInterval`. When `periodInterval` is 0, `stride` causes a crash since it cannot be created with the step's value being 0.

This PR fixes the crash by handling potential issues in two places:
- Decoder: we were using 0 for the default value for `periodInterval`. This was fixed by setting the value to 1 for parity with the [default empty product subscription](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/Extensions/ProductSubscription%2BEmpty.swift).
- Also decoder: a potential cause for the failure of the decoder is that sites using plugins that alter the response of products. So instead of returning String for numbers like in the documentation, the response returns integers or doubles. I've updated the decoder to accept alternative types to avoid failures.
- `SubscriptionExpiryViewModel`: I added a check to ensure `periodInterval` is larger than 0 before proceeding to `stride`. If the value is 0, the selected length and length options are "Never expires".

## Testing instructions

Prerequisite: make sure there is an existing product subscription on your store.

We don't know the exact plugins that can cause this behavior, but you can test using a tool like Charles Proxy or Proxyman to intercept and modify the response:

- Set a breakpoint in your tool of choice for requests to the /wc/v3/products endpoint.
- Build and run the app.
- Open the Products tab. 
- Modify the response to update the numeric values in `meta` from strings to integers for a subscription product (check `product-subscription-alternative-types.json` if you need an example).
- Execute the response.
- Select the updated subscription product. Confirm that all the subscription details are parsed correctly on the Price, Free Trial, Shipping, Expire after rows.

## Screenshots

N/A

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Subscription/SubscriptionExpiryViewModelTests.swift [hunk 1]
```diff
@@ -15,6 +15,16 @@ final class SubscriptionExpiryViewModelTests: XCTestCase {
...
 
+    func test_selectedLength_and_lengthOptions_are_correct_if_periodInterval_is_0() {
+        // Given
+        let subscription = ProductSubscription.fake().copy(periodInterval: "0")
+        let viewModel = SubscriptionExpiryViewModel(subscription: subscription) { _, _ in }
+
+        // Then
+        XCTAssertEqual(viewModel.selectedLength.value, 0)
+        XCTAssertEqual(viewModel.lengthOptions.count, 1)
+    }
```
Comment (iamgabrielma) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Subscription/SubscriptionExpiryViewModelTests.swift:26]: Question: Should we add an additional test to check for negative numbers as well, just for safety? Currently will work nicely because of the `periodInterval > 0` check, but if we happen to change the implementation for `periodInterval != 0` for some reason, this test would still pass but a negative number into the stride would also cause a crash.

**Review by iamgabrielma — APPROVED** [review]
LGTM :shipit: 

- For your consideration, since is not specifically from this PR: Should we move the _"Configure length options based on billing period and interval"_ logic from init to a separate configuration method?
- Failing UI Test: I'm unable to replicate the case for the failing UI test on iPad. I've seen this one and others it failing sometimes due to a non-dismissed keychain notification which causes the expected operations to time-out, but it's entirely unrelated to your PR:

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Subscription/SubscriptionExpiryViewModelTests.swift (hunk 1 above)
Comment (itsmeichigo) [reply WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Subscription/SubscriptionExpiryViewModelTests.swift:26]: Thanks for the suggestion! We don't expect period interval to be less than 0, and it's unlikely that a plugin would make this happen, so I figure it's not necessary to cover that case.

**Comment by itsmeichigo** [comment]
Thank you @iamgabrielma for the review and suggestions!

> For your consideration, since is not specifically from this PR: Should we move the "Configure length options based on billing period and interval" logic from init to a separate configuration method?

This was an implementation decision from @selanthiraiyan, personally, I don't find much problem with it. I'd keep the implementation as-is to avoid making too many changes for a beta fix.

> Failing UI Test: I'm unable to replicate the case for the failing UI test on iPad. I've seen this one and others it failing sometimes due to a non-dismissed keychain notification which causes the expected operations to time-out, but it's entirely unrelated to your PR:

This is a tricky UI test issue - I could not reproduce it locally with iOS 17 simulator. I wonder if it's a problem with iOS 16 considering that our CI hasn't been updated yet. Tagging @jostnes for awareness.

I've retried the UI test - but if it still fails I'll force merge the PR to get the fix in time for v16.6.

**Comment by jostnes** [comment]
👋 @itsmeichigo! thanks for the ping, i had a look, and the failing test in the [first build](https://buildkite.com/automattic/woocommerce-ios/builds/18107) looks familiar. i believe that should be fixed by https://github.com/woocommerce/woocommerce-ios/pull/11418 (merged a couple of days ago to `trunk` but not in `release/16.6`)

should be safe to force merge or to see it 🟢 feel free to get the commit into this PR for the failure to go away. (also your assessment is correct; it's a tricky UI test issue not reproducible locally but somehow reproducible quite easily in CI so could be related to CI setup)

**Comment by itsmeichigo** [comment]
cc @spencertransier Please include this fix in the final build for 16.6. Thank you 🙏
